packages/@aws-cdk/aws-iam/lib/policy-document.ts

```diff
@@ -39,8 +42,22 @@ export class PolicyDocument extends cdk.Token {
     return this.statements.length;
   }
 
+  /**
+   * Adds a statement to the policy document. Prevents adding duplicate
+   * statements.
+   *
+   * Removal of duplicate statements containing tokens resolving to the same
+   * values but represented by different strings is currently not supported.
+   * This is suboptimal but safe.
+   *
+   * @param statement the statement to add.
+   */
   public addStatement(statement: PolicyStatement): PolicyDocument {
-    this.statements.push(statement);
```

Comment:
Instead of doing this, implement `IResolvedValuePostProcessor` and do it in the postProcess method.

This is guaranteed to receive the post-resolution Statement, and you can do safe pairwise comparisons based on `JSON.stringify()`.